File: python/143_Reorder_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head:
            return
        
        slow = fast = head
        while(fast and fast.next):
            slow, fast = slow.next, fast.next.next
        
        prev, cur = None, slow
        while cur:
            cur.next, prev, cur = prev, cur, cur.next
        
        l1, l2 = head, prev
        while l2.next:
            l1.next, l1 = l2, l1.next
            l2.next, l2 = l1, l2.next

File: java/143_Reorder_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null){
            return;
        }
        if (head.next == null){
            return;
        }

        ListNode mid = findMiddle(head);
        ListNode l1 = head;
        ListNode l2 = mid.next;
        mid.next = null;
        
        l2 = reverseList(l2);
        mergeList(l1,l2);
    }
    
    public ListNode findMiddle(ListNode head){
        ListNode slow = head;
        ListNode fast = head;
        while(fast.next!=null && fast.next.next!=null){
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }
    
    public ListNode reverseList(ListNode head){
        ListNode prev = null;
        ListNode cur = head;
        ListNode nxt;
        while(cur!=null){
            nxt = cur.next;
            cur.next = prev;
            prev = cur;
            cur = nxt;
        }
        return prev;
    }
    
    public void mergeList(ListNode first,ListNode second){
        ListNode first_temp;
        ListNode second_temp;
        while(first != null && second != null){
            first_temp = first.next;
            second_temp = second.next;
            first.next = second;
            second.next = first_temp;
            
            first = first_temp;
            second = second_temp;
        }
    }
}